### PR TITLE
v1.0.0

### LINKED ISSUE #6: Enhance performance (at least no worse than Java randomUUID)

### PR DESCRIPTION
**Pull request by zelark**
# nano-id 1.0.0

## Breaking changes
All breaking changes touch only `nano-id.custom`, so if you haven't used it, there is nothing to worry about. The first thing to note is that `nano-id.custom/generate` has gone, instead of it there is a new function called `custom` which lives now in `nano-id.core` ns. The second thing is that `custom` got a new signatures: `[alphabet size]` and `[alphabet size random]`. Once you created a generator you can't change `size`, now it's closured. Check out [updated examples](https://github.com/zelark/nano-id#custom-id-generator).

## Improvements
Now the default implementation goes even faster, it's almost as fast as UUID: 1.72µs and 1.60µs respectively. Custom also got a bit faster.

Closes #6.

### CODE REVIEW

File: src/nano_id/core.cljc [hunk 1]
```diff
@@ -1,28 +1,54 @@
...
+      :cljs (let [mask 0x3f]
+              (loop [bytes (rnd/random-bytes size)
+                     id ""]
+                   (if bytes
+                     (recur (next bytes)
+                            (->> (first bytes)
+                                 (bit-and mask)
+                                 alphabet
+                                 (str id)))
+                     (str id)))))))
```
Comment (dosbol) [inline src/nano_id/core.cljc]:
```suggestion
                     id))))))
```

Comment (zelark) [reply src/nano_id/core.cljc]: Nice catch, forgot about it when separated clj and cljs versions.

File: src/nano_id/NanoID.java [hunk 2]
```diff
@@ -25,4 +26,20 @@ public static String nanoID(int size) {
...
+  
+  public static String custom(char[] alphabet, int size, IFn random, int step, int mask) {
+    final StringBuilder id = new StringBuilder();
+    while (true) {
+      final byte[] bytes = (byte[]) random.invoke(step);
+      for (int i = 0; i < step; ++i) {
+      final int index = bytes[i] & mask;
+      if (index < alphabet.length) {
+          id.append(alphabet[index]);
+          if (id.length() == size) {
```
Comment (dosbol) [inline src/nano_id/NanoID.java:38]: what if size = 0 ?

File: src/nano_id/core.cljc [hunk 3]
```diff
@@ -28,27 +30,29 @@
...
-  Also you can provide your own random bytes generator."
+  Also you can provide your own random bytes generator. It should be a function
+  that takes one argument - number of bytes, and returns a byte array."
   ([alphabet size]
    (custom alphabet size rnd/random-bytes))
   ([alphabet size random]
    (assert (<= 2 (count alphabet) 256) "alphabet must contain from 2 to 256 characters.")
    (assert (pos? size) "size must be positive.")
-   (let [alphabet      (mapv str alphabet)
+   (let [alphabet      #?(:clj (.toCharArray ^String alphabet) :cljs (mapv str alphabet))
```
Comment (dosbol) [inline src/nano_id/core.cljc:40]:
```suggestion
   (let [alphabet      #?(:clj (.toCharArray ^String alphabet) :cljs alphabet)
```

**Review by zelark — COMMENTED** [review]
Checked.

**Pull request merged**